File: ts/src/lexer.ts
import { HronError, type Span } from "./error.js";

export interface Token {
  kind: TokenKind;
  span: Span;
}

export type TokenKind =
  | { type: "every" }
  | { type: "on" }
  | { type: "at" }
  | { type: "from" }
  | { type: "to" }
  | { type: "in" }
  | { type: "of" }
  | { type: "the" }
  | { type: "last" }
  | { type: "except" }
  | { type: "until" }
  | { type: "starting" }
  | { type: "during" }
  | { type: "year" }
  | { type: "day" }
  | { type: "weekday" }
  | { type: "weekend" }
  | { type: "weeks" }
  | { type: "month" }
  | { type: "dayName"; name: string }
  | { type: "monthName"; name: string }
  | { type: "ordinal"; name: string }
  | { type: "intervalUnit"; unit: string }

  | { type: "number"; value: number }
  | { type: "ordinalNumber"; value: number }
  | { type: "time"; hour: number; minute: number }
  | { type: "isoDate"; date: string }
  | { type: "comma" }
  | { type: "timezone"; tz: string };

export function tokenize(input: string): Token[] {
  const lexer = new Lexer(input);
  return lexer.tokenize();
}

class Lexer {
  private input: string;
  private pos: number;
  private afterIn: boolean;

  constructor(input: string) {
    this.input = input;
    this.pos = 0;
    this.afterIn = false;
  }

  tokenize(): Token[] {
    const tokens: Token[] = [];
    while (true) {
      this.skipWhitespace();
      if (this.pos >= this.input.length) break;

      if (this.afterIn) {
        this.afterIn = false;
        tokens.push(this.lexTimezone());
        continue;
      }

      const start = this.pos;
      const ch = this.input[this.pos];

      if (ch === ",") {
        this.pos++;
        tokens.push({ kind: { type: "comma" }, span: { start, end: this.pos } });
        continue;
      }

      if (isDigit(ch)) {
        tokens.push(this.lexNumberOrTimeOrDate());
        continue;
      }

      if (isAlpha(ch)) {
        tokens.push(this.lexWord());
        continue;
      }

      throw HronError.lex(
        `unexpected character '${ch}'`,
        { start, end: start + 1 },
        this.input,
      );
    }
    return tokens;
  }

  private skipWhitespace(): void {
    while (this.pos < this.input.length && isWhitespace(this.input[this.pos])) {
      this.pos++;
    }
  }

  private lexTimezone(): Token {
    this.skipWhitespace();
    const start = this.pos;
    while (
      this.pos < this.input.length &&
      !isWhitespace(this.input[this.pos])
    ) {
      this.pos++;
    }
    const tz = this.input.slice(start, this.pos);
    if (tz.length === 0) {
      throw HronError.lex(
        "expected timezone after 'in'",
        { start, end: start + 1 },
        this.input,
      );
    }
    return { kind: { type: "timezone", tz }, span: { start, end: this.pos } };
  }

  private lexNumberOrTimeOrDate(): Token {
    const start = this.pos;
    const numStart = this.pos;
    while (this.pos < this.input.length && isDigit(this.input[this.pos])) {
      this.pos++;
    }
    const digits = this.input.slice(numStart, this.pos);

    // Check for ISO date: YYYY-MM-DD
    if (
      digits.length === 4 &&
      this.pos < this.input.length &&
      this.input[this.pos] === "-"
    ) {
      const remaining = this.input.slice(start);
      if (
        remaining.length >= 10 &&
        remaining[4] === "-" &&
        isDigit(remaining[5]) &&
        isDigit(remaining[6]) &&
        remaining[7] === "-" &&
        isDigit(remaining[8]) &&
        isDigit(remaining[9])
      ) {
        this.pos = start + 10;
        return {
          kind: { type: "isoDate", date: this.input.slice(start, this.pos) },
          span: { start, end: this.pos },
        };
      }
    }

    // Check for time: HH:MM
    if (
      (digits.length === 1 || digits.length === 2) &&
      this.pos < this.input.length &&
      this.input[this.pos] === ":"
    ) {
      this.pos++; // skip ':'
      const minStart = this.pos;
      while (this.pos < this.input.length && isDigit(this.input[this.pos])) {
        this.pos++;
      }
      const minDigits = this.input.slice(minStart, this.pos);
      if (minDigits.length === 2) {
        const hour = parseInt(digits, 10);
        const minute = parseInt(minDigits, 10);
        if (hour > 23 || minute > 59) {
          throw HronError.lex(
            "invalid time",
            { start, end: this.pos },
            this.input,
          );
        }
        return {
          kind: { type: "time", hour, minute },
          span: { start, end: this.pos },
        };
      }
    }

    const num = parseInt(digits, 10);
    if (isNaN(num)) {
      throw HronError.lex(
        "invalid number",
        { start, end: this.pos },
        this.input,
      );
    }

    // Check for ordinal suffix: st, nd, rd, th
    if (this.pos + 1 < this.input.length) {
      const suffix = this.input.slice(this.pos, this.pos + 2).toLowerCase();
      if (suffix === "st" || suffix === "nd" || suffix === "rd" || suffix === "th") {
        this.pos += 2;
        return {
          kind: { type: "ordinalNumber", value: num },
          span: { start, end: this.pos },
        };
      }
    }

    return {
      kind: { type: "number", value: num },
      span: { start, end: this.pos },
    };
  }

  private lexWord(): Token {
    const start = this.pos;
    while (
      this.pos < this.input.length &&
      (isAlphanumeric(this.input[this.pos]) || this.input[this.pos] === "_")
    ) {
      this.pos++;
    }
    const word = this.input.slice(start, this.pos).toLowerCase();
    const span = { start, end: this.pos };

    const kind = KEYWORD_MAP[word];
    if (kind === undefined) {
      throw HronError.lex(`unknown keyword '${word}'`, span, this.input);
    }

    if (kind.type === "in") {
      this.afterIn = true;
    }

    return { kind, span };
  }
}

const KEYWORD_MAP: Record<string, TokenKind> = {
  every: { type: "every" },
  on: { type: "on" },
  at: { type: "at" },
  from: { type: "from" },
  to: { type: "to" },
  in: { type: "in" },
  of: { type: "of" },
  the: { type: "the" },
  last: { type: "last" },
  except: { type: "except" },
  until: { type: "until" },
  starting: { type: "starting" },
  during: { type: "during" },
  year: { type: "year" },

  day: { type: "day" },
  weekday: { type: "weekday" },
  weekdays: { type: "weekday" },
  weekend: { type: "weekend" },
  weekends: { type: "weekend" },
  weeks: { type: "weeks" },
  week: { type: "weeks" },
  month: { type: "month" },

  monday: { type: "dayName", name: "monday" },
  mon: { type: "dayName", name: "monday" },
  tuesday: { type: "dayName", name: "tuesday" },
  tue: { type: "dayName", name: "tuesday" },
  wednesday: { type: "dayName", name: "wednesday" },
  wed: { type: "dayName", name: "wednesday" },
  thursday: { type: "dayName", name: "thursday" },
  thu: { type: "dayName", name: "thursday" },
  friday: { type: "dayName", name: "friday" },
  fri: { type: "dayName", name: "friday" },
  saturday: { type: "dayName", name: "saturday" },
  sat: { type: "dayName", name: "saturday" },
  sunday: { type: "dayName", name: "sunday" },
  sun: { type: "dayName", name: "sunday" },

  january: { type: "monthName", name: "jan" },
  jan: { type: "monthName", name: "jan" },
  february: { type: "monthName", name: "feb" },
  feb: { type: "monthName", name: "feb" },
  march: { type: "monthName", name: "mar" },
  mar: { type: "monthName", name: "mar" },
  april: { type: "monthName", name: "apr" },
  apr: { type: "monthName", name: "apr" },
  may: { type: "monthName", name: "may" },
  june: { type: "monthName", name: "jun" },
  jun: { type: "monthName", name: "jun" },
  july: { type: "monthName", name: "jul" },
  jul: { type: "monthName", name: "jul" },
  august: { type: "monthName", name: "aug" },
  aug: { type: "monthName", name: "aug" },
  september: { type: "monthName", name: "sep" },
  sep: { type: "monthName", name: "sep" },
  october: { type: "monthName", name: "oct" },
  oct: { type: "monthName", name: "oct" },
  november: { type: "monthName", name: "nov" },
  nov: { type: "monthName", name: "nov" },
  december: { type: "monthName", name: "dec" },
  dec: { type: "monthName", name: "dec" },

  first: { type: "ordinal", name: "first" },
  second: { type: "ordinal", name: "second" },
  third: { type: "ordinal", name: "third" },
  fourth: { type: "ordinal", name: "fourth" },
  fifth: { type: "ordinal", name: "fifth" },

  min: { type: "intervalUnit", unit: "min" },
  mins: { type: "intervalUnit", unit: "min" },
  minute: { type: "intervalUnit", unit: "min" },
  minutes: { type: "intervalUnit", unit: "min" },
  hour: { type: "intervalUnit", unit: "hours" },
  hours: { type: "intervalUnit", unit: "hours" },
  hr: { type: "intervalUnit", unit: "hours" },
  hrs: { type: "intervalUnit", unit: "hours" },
};

function isDigit(ch: string): boolean {
  return ch >= "0" && ch <= "9";
}

function isAlpha(ch: string): boolean {
  return (ch >= "a" && ch <= "z") || (ch >= "A" && ch <= "Z");
}

function isAlphanumeric(ch: string): boolean {
  return isDigit(ch) || isAlpha(ch);
}

function isWhitespace(ch: string): boolean {
  return ch === " " || ch === "\t" || ch === "\n" || ch === "\r";
}

File: dart/lib/src/lexer.dart
import 'ast.dart';
import 'error.dart';

// --- Token kinds ---

sealed class TokenKind {}

class EveryToken extends TokenKind {}

class OnToken extends TokenKind {}

class AtToken extends TokenKind {}

class FromToken extends TokenKind {}

class ToToken extends TokenKind {}

class InToken extends TokenKind {}

class OfToken extends TokenKind {}

class TheToken extends TokenKind {}

class LastToken extends TokenKind {}

class ExceptToken extends TokenKind {}

class UntilToken extends TokenKind {}

class StartingToken extends TokenKind {}

class DuringToken extends TokenKind {}

class NearestToken extends TokenKind {}

class NextToken extends TokenKind {}

class PreviousToken extends TokenKind {}

class YearToken extends TokenKind {}

class DayToken extends TokenKind {}

class WeekdayKeyToken extends TokenKind {}

class WeekendKeyToken extends TokenKind {}

class WeeksToken extends TokenKind {}

class MonthToken extends TokenKind {}

class CommaToken extends TokenKind {}

class DayNameToken extends TokenKind {
  final Weekday name;
  DayNameToken(this.name);
}

class MonthNameToken extends TokenKind {
  final MonthName name;
  MonthNameToken(this.name);
}

class OrdinalToken extends TokenKind {
  final OrdinalPosition name;
  OrdinalToken(this.name);
}

class IntervalUnitToken extends TokenKind {
  final IntervalUnit unit;
  IntervalUnitToken(this.unit);
}

class NumberToken extends TokenKind {
  final int value;
  NumberToken(this.value);
}

class OrdinalNumberToken extends TokenKind {
  final int value;
  OrdinalNumberToken(this.value);
}

class TimeToken extends TokenKind {
  final int hour;
  final int minute;
  TimeToken(this.hour, this.minute);
}

class IsoDateToken extends TokenKind {
  final String date;
  IsoDateToken(this.date);
}

class TimezoneToken extends TokenKind {
  final String tz;
  TimezoneToken(this.tz);
}

// --- Token ---

class Token {
  final TokenKind kind;
  final Span span;
  Token(this.kind, this.span);
}

// --- Tokenizer ---

List<Token> tokenize(String input) => _Lexer(input).tokenize();

class _Lexer {
  final String input;
  int pos = 0;
  bool afterIn = false;

  _Lexer(this.input);

  List<Token> tokenize() {
    final tokens = <Token>[];
    while (true) {
      _skipWhitespace();
      if (pos >= input.length) break;

      if (afterIn) {
        afterIn = false;
        tokens.add(_lexTimezone());
        continue;
      }

      final start = pos;
      final ch = input[pos];

      if (ch == ',') {
        pos++;
        tokens.add(Token(CommaToken(), Span(start, pos)));
        continue;
      }

      if (_isDigit(ch)) {
        tokens.add(_lexNumberOrTimeOrDate());
        continue;
      }

      if (_isAlpha(ch)) {
        tokens.add(_lexWord());
        continue;
      }

      throw HronError.lex(
        "unexpected character '$ch'",
        Span(start, start + 1),
        input,
      );
    }
    return tokens;
  }

  void _skipWhitespace() {
    while (pos < input.length && _isWhitespace(input[pos])) {
      pos++;
    }
  }

  Token _lexTimezone() {
    _skipWhitespace();
    final start = pos;
    while (pos < input.length && !_isWhitespace(input[pos])) {
      pos++;
    }
    final tz = input.substring(start, pos);
    if (tz.isEmpty) {
      throw HronError.lex(
        "expected timezone after 'in'",
        Span(start, start + 1),
        input,
      );
    }
    return Token(TimezoneToken(tz), Span(start, pos));
  }

  Token _lexNumberOrTimeOrDate() {
    final start = pos;
    final numStart = pos;
    while (pos < input.length && _isDigit(input[pos])) {
      pos++;
    }
    final digits = input.substring(numStart, pos);

    // ISO date: YYYY-MM-DD
    if (digits.length == 4 && pos < input.length && input[pos] == '-') {
      final remaining = input.substring(start);
      if (remaining.length >= 10 &&
          remaining[4] == '-' &&
          _isDigit(remaining[5]) &&
          _isDigit(remaining[6]) &&
          remaining[7] == '-' &&
          _isDigit(remaining[8]) &&
          _isDigit(remaining[9])) {
        pos = start + 10;
        return Token(
          IsoDateToken(input.substring(start, pos)),
          Span(start, pos),
        );
      }
    }

    // Time: HH:MM
    if ((digits.length == 1 || digits.length == 2) &&
        pos < input.length &&
        input[pos] == ':') {
      pos++; // skip ':'
      final minStart = pos;
      while (pos < input.length && _isDigit(input[pos])) {
        pos++;
      }
      final minDigits = input.substring(minStart, pos);
      if (minDigits.length == 2) {
        final hour = int.parse(digits);
        final minute = int.parse(minDigits);
        if (hour > 23 || minute > 59) {
          throw HronError.lex('invalid time', Span(start, pos), input);
        }
        return Token(TimeToken(hour, minute), Span(start, pos));
      }
    }

    final num = int.tryParse(digits);
    if (num == null) {
      throw HronError.lex('invalid number', Span(start, pos), input);
    }

    // Ordinal suffix: st, nd, rd, th
    if (pos + 1 < input.length) {
      final suffix = input.substring(pos, pos + 2).toLowerCase();
      if (suffix == 'st' ||
          suffix == 'nd' ||
          suffix == 'rd' ||
          suffix == 'th') {
        pos += 2;
        return Token(OrdinalNumberToken(num), Span(start, pos));
      }
    }

    return Token(NumberToken(num), Span(start, pos));
  }

  Token _lexWord() {
    final start = pos;
    while (pos < input.length &&
        (_isAlphanumeric(input[pos]) || input[pos] == '_')) {
      pos++;
    }
    final word = input.substring(start, pos).toLowerCase();
    final span = Span(start, pos);

    final kind = _keywordMap[word];
    if (kind == null) {
      throw HronError.lex("unknown keyword '$word'", span, input);
    }

    if (kind is InToken) {
      afterIn = true;
    }

    return Token(kind, span);
  }
}

String tokenKindType(TokenKind kind) => switch (kind) {
  EveryToken() => 'every',
  OnToken() => 'on',
  AtToken() => 'at',
  FromToken() => 'from',
  ToToken() => 'to',
  InToken() => 'in',
  OfToken() => 'of',
  TheToken() => 'the',
  LastToken() => 'last',
  ExceptToken() => 'except',
  UntilToken() => 'until',
  StartingToken() => 'starting',
  DuringToken() => 'during',
  NearestToken() => 'nearest',
  NextToken() => 'next',
  PreviousToken() => 'previous',
  YearToken() => 'year',
  DayToken() => 'day',
  WeekdayKeyToken() => 'weekday',
  WeekendKeyToken() => 'weekend',
  WeeksToken() => 'weeks',
  MonthToken() => 'month',
  CommaToken() => 'comma',
  DayNameToken() => 'dayName',
  MonthNameToken() => 'monthName',
  OrdinalToken() => 'ordinal',
  IntervalUnitToken() => 'intervalUnit',
  NumberToken() => 'number',
  OrdinalNumberToken() => 'ordinalNumber',
  TimeToken() => 'time',
  IsoDateToken() => 'isoDate',
  TimezoneToken() => 'timezone',
};

final _keywordMap = <String, TokenKind>{
  'every': EveryToken(),
  'on': OnToken(),
  'at': AtToken(),
  'from': FromToken(),
  'to': ToToken(),
  'in': InToken(),
  'of': OfToken(),
  'the': TheToken(),
  'last': LastToken(),
  'except': ExceptToken(),
  'until': UntilToken(),
  'starting': StartingToken(),
  'during': DuringToken(),
  'nearest': NearestToken(),
  'next': NextToken(),
  'previous': PreviousToken(),
  'year': YearToken(),
  'years': YearToken(),
  'day': DayToken(),
  'days': DayToken(),
  'weekday': WeekdayKeyToken(),
  'weekdays': WeekdayKeyToken(),
  'weekend': WeekendKeyToken(),
  'weekends': WeekendKeyToken(),
  'weeks': WeeksToken(),
  'week': WeeksToken(),
  'month': MonthToken(),
  'months': MonthToken(),
  'monday': DayNameToken(Weekday.monday),
  'mon': DayNameToken(Weekday.monday),
  'tuesday': DayNameToken(Weekday.tuesday),
  'tue': DayNameToken(Weekday.tuesday),
  'wednesday': DayNameToken(Weekday.wednesday),
  'wed': DayNameToken(Weekday.wednesday),
  'thursday': DayNameToken(Weekday.thursday),
  'thu': DayNameToken(Weekday.thursday),
  'friday': DayNameToken(Weekday.friday),
  'fri': DayNameToken(Weekday.friday),
  'saturday': DayNameToken(Weekday.saturday),
  'sat': DayNameToken(Weekday.saturday),
  'sunday': DayNameToken(Weekday.sunday),
  'sun': DayNameToken(Weekday.sunday),
  'january': MonthNameToken(MonthName.jan),
  'jan': MonthNameToken(MonthName.jan),
  'february': MonthNameToken(MonthName.feb),
  'feb': MonthNameToken(MonthName.feb),
  'march': MonthNameToken(MonthName.mar),
  'mar': MonthNameToken(MonthName.mar),
  'april': MonthNameToken(MonthName.apr),
  'apr': MonthNameToken(MonthName.apr),
  'may': MonthNameToken(MonthName.may),
  'june': MonthNameToken(MonthName.jun),
  'jun': MonthNameToken(MonthName.jun),
  'july': MonthNameToken(MonthName.jul),
  'jul': MonthNameToken(MonthName.jul),
  'august': MonthNameToken(MonthName.aug),
  'aug': MonthNameToken(MonthName.aug),
  'september': MonthNameToken(MonthName.sep),
  'sep': MonthNameToken(MonthName.sep),
  'october': MonthNameToken(MonthName.oct),
  'oct': MonthNameToken(MonthName.oct),
  'november': MonthNameToken(MonthName.nov),
  'nov': MonthNameToken(MonthName.nov),
  'december': MonthNameToken(MonthName.dec),
  'dec': MonthNameToken(MonthName.dec),
  'first': OrdinalToken(OrdinalPosition.first),
  'second': OrdinalToken(OrdinalPosition.second),
  'third': OrdinalToken(OrdinalPosition.third),
  'fourth': OrdinalToken(OrdinalPosition.fourth),
  'fifth': OrdinalToken(OrdinalPosition.fifth),
  'min': IntervalUnitToken(IntervalUnit.min),
  'mins': IntervalUnitToken(IntervalUnit.min),
  'minute': IntervalUnitToken(IntervalUnit.min),
  'minutes': IntervalUnitToken(IntervalUnit.min),
  'hour': IntervalUnitToken(IntervalUnit.hours),
  'hours': IntervalUnitToken(IntervalUnit.hours),
  'hr': IntervalUnitToken(IntervalUnit.hours),
  'hrs': IntervalUnitToken(IntervalUnit.hours),
};

bool _isDigit(String ch) => ch.codeUnitAt(0) >= 48 && ch.codeUnitAt(0) <= 57;

bool _isAlpha(String ch) {
  final c = ch.codeUnitAt(0);
  return (c >= 65 && c <= 90) || (c >= 97 && c <= 122);
}

bool _isAlphanumeric(String ch) => _isDigit(ch) || _isAlpha(ch);

bool _isWhitespace(String ch) =>
    ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r';
